File: packages/agents/src/optimizers/generation/optimize-generation.ts
import {
  AxAI,
  AxAIGoogleGeminiModel,
  type AxMetricFn,
  AxDefaultCostTracker,
  AxCheckpointLoadFn,
} from '@ax-llm/ax';
import {
  generationDataset,
  type GenerationExample,
} from '../datasets/generation.dataset';
import { generationProgram } from '../../core/programs/generation.program';
import { getGeminiApiKey, logger } from '../..';
import { generationMetricFn } from './utils';
import { AxMiPRO } from '@ax-llm/ax';
import * as fs from 'fs';
import * as path from 'path';

// Setup student AI model
const apiKey = getGeminiApiKey();
const studentAI = new AxAI({
  name: 'google-gemini',
  apiKey,
  models: [
    {
      key: 'gemini-fast',
      model: AxAIGoogleGeminiModel.Gemini25Flash,
      description: 'Fast model for simple tasks',
    },
  ],
});

const costTracker = new AxDefaultCostTracker({
  maxTokens: 300000,
  maxCost: 1.0,
});

// Checkpoint functions
const checkpointSave = async (state: any) => {
  const id = `checkpoint_${Date.now()}`;
  try {
    fs.writeFileSync(`${id}.json`, JSON.stringify(state, null, 2));
    logger.info(`Checkpoint saved to ${id}.json`);
  } catch (error) {
    logger.error('Failed to save checkpoint:', error);
  }
  return id;
};

const checkpointLoad: AxCheckpointLoadFn = async (id) => {
  try {
    const data = fs.readFileSync(`${id}.json`, 'utf8');
    logger.info(`Checkpoint loaded successfully from ${id}.json`);
    return JSON.parse(data);
  } catch (error) {
    logger.error('Failed to load checkpoint:', error);
    return null;
  }
};

// Optimizer instance
const optimizer = new AxMiPRO({
  studentAI,
  // @ts-ignore
  examples: generationDataset,
  targetScore: 0.9,
  verbose: true,
  costTracker,
  checkpointSave,
  checkpointLoad,
  checkpointInterval: 10,
  options: {
    maxBootstrappedDemos: 15,
    maxLabeledDemos: 10,
    numTrials: 8,
  },
});

/**
 * Evaluates baseline performance on a subset of the dataset.
 */
const evaluateBaseline = async () => {
  console.log('Evaluating baseline performance...');
  let totalScore = 0;

  for (let i = 0; i < Math.min(5, generationDataset.length); i++) {
    const example = generationDataset[i];
    const modelKey = 'gemini-fast';

    try {
      const result = await generationProgram.forward(
        studentAI,
        {
          chat_history: example.chat_history || '',
          query: example.query,
          context: example.context,
        },
        { model: modelKey },
      );

      const score = await generationMetricFn({
        prediction: result,
        // @ts-ignore
        example,
      });

      totalScore += score;
    } catch (error) {
      console.error(`Example ${i + 1}: Error - ${error}`);
    }
  }

  const avgScore = totalScore / Math.min(5, generationDataset.length);
  console.log(`Baseline average score: ${avgScore.toFixed(4)}`);
  return avgScore;
};

/**
 * Main function to run the optimization.
 */
const main = async () => {
  console.log('Starting generation program optimization...');

  const baselineScore = await evaluateBaseline();

  const startTime = Date.now();
  const result = await optimizer.compile(generationProgram, generationMetricFn);
  const duration = (Date.now() - startTime) / 1000;

  console.log('Optimization Complete!');
  console.log(`Duration: ${duration.toFixed(2)}s`);
  console.log(`Baseline Score: ${baselineScore.toFixed(4)}`);
  console.log(`Best Score: ${result.bestScore.toFixed(4)}`);
  console.log(
    `Improvement: ${((result.bestScore - baselineScore) * 100).toFixed(1)}%`,
  );
  console.log(`Total API Calls: ${result.stats.totalCalls}`);
  console.log(`Demos Generated: ${result.demos?.length || 0}`);

  if (result.demos) {
    const outputPath = path.join(__dirname, 'optimized-generation-demos.json');
    fs.writeFileSync(outputPath, JSON.stringify(result.demos, null, 2));
    console.log(`Demos saved to: ${outputPath}`);
  }

  const programConfig = JSON.stringify(result.finalConfiguration, null, 2);
  fs.writeFileSync(
    path.join(__dirname, 'optimized-config.json'),
    programConfig,
  );

  console.log(`Cost: $${costTracker.getCurrentCost().toFixed(4)}`);
  console.log(`Tokens: ${JSON.stringify(costTracker.getTokenUsage())}`);
};

main().catch((error) => {
  console.error('Optimization failed:', error);
  process.exit(1);
});

File: python/src/cairo_coder/optimizers/generation/optimize_generation.py
#!/usr/bin/env python3
"""Script to optimize the generation program using DSPy and the Starklings dataset."""

import json
import time
from pathlib import Path
from typing import List

import dspy
import structlog
from dspy import MIPROv2

from cairo_coder.dspy.generation_program import GenerationProgram
from cairo_coder.optimizers.generation.utils import generation_metric

logger = structlog.get_logger(__name__)


def load_dataset(dataset_path: str) -> List[dspy.Example]:
    """Load dataset from JSON file."""
    with open(dataset_path, "r", encoding="utf-8") as f:
        data = json.load(f)

    examples = []
    for ex in data["examples"]:
        example = dspy.Example(
            query=ex["query"],
            chat_history=ex["chat_history"],
            context=ex["context"],
            expected=ex["expected"],
        ).with_inputs("query", "chat_history", "context")
        examples.append(example)

    logger.info("Loaded dataset", count=len(examples), examples=examples)
    return examples

def evaluate_baseline(examples: List[dspy.Example]) -> float:
    """Evaluate baseline performance on first 5 examples."""
    logger.info("Evaluating baseline performance")

    program = GenerationProgram()
    scores = []

    for i, example in enumerate(examples[:5]):
        try:
            prediction = program.forward(
                query=example.query,
                chat_history=example.chat_history,
                context=example.context,
            )
            score = generation_metric(example, prediction)
            scores.append(score)
            logger.debug(
                "Baseline evaluation",
                example=i,
                score=score,
                query=example.query[:50] + "...",
            )
        except Exception as e:
            logger.error("Error in baseline evaluation", example=i, error=str(e))
            scores.append(0.0)

    avg_score = sum(scores) / len(scores) if scores else 0.0
    logger.info("Baseline evaluation complete", average_score=avg_score)
    return avg_score

def run_optimization(trainset: List[dspy.Example], valset: List[dspy.Example]) -> tuple:
    """Run the optimization process using MIPROv2."""
    logger.info("Starting optimization process")

    # Initialize program
    program = GenerationProgram()

    # Configure optimizer
    optimizer = MIPROv2(
        metric=generation_metric,
        auto="light",
        max_bootstrapped_demos=4,
        max_labeled_demos=4,
    )

    # Run optimization
    start_time = time.time()
    optimized_program = optimizer.compile(
        program,
        trainset=trainset,
        valset=valset,  # Use trainset for validation
    )
    duration = time.time() - start_time

    logger.info(
        "Optimization completed",
        duration=f"{duration:.2f}s",
    )

    return optimized_program, duration

def main():
    """Main optimization workflow."""
    logger.info("Starting generation program optimization")

    # Setup DSPy
    lm = dspy.LM("gemini/gemini-2.5-flash", max_tokens=30000)
    dspy.settings.configure(lm=lm)
    logger.info("Configured DSPy with Gemini 2.5 Flash")

    # Load dataset
    dataset_path = "optimizers/datasets/generation_dataset.json"
    if not Path(dataset_path).exists():
        logger.error("Dataset not found. Please run generate_starklings_dataset.py first.")
        return

    examples = load_dataset(dataset_path)

    # Split dataset (70/30 for train/val)
    split_idx = int(0.7* len(examples))
    trainset = examples
    trainset = examples[:split_idx]
    valset = examples[split_idx:]

    logger.info(
        "Dataset split",
        train_size=len(trainset),
        val_size=len(valset),
        total=len(examples),
    )

    # Evaluate baseline
    baseline_score = evaluate_baseline(trainset)

    # Run optimization
    optimized_program, duration = run_optimization(trainset, valset)

    # Save optimized program
    optimized_program.save("optimizers/results/optimized_generation_program.json")
    logger.info("Optimization complete. Results saved to optimizers/results/")



    # Evaluate final performance
    final_scores = []
    for example in valset:  # Test on validation set
        try:
            prediction = optimized_program.forward(
                query=example.query,
                chat_history=example.chat_history,
                context=example.context,
            )
            score = generation_metric(example, prediction)
            final_scores.append(score)
        except Exception as e:
            logger.error("Error in final evaluation", error=str(e))
            final_scores.append(0.0)

    final_score = sum(final_scores) / len(final_scores) if final_scores else 0.0
    improvement = final_score - baseline_score

    # Calculate costs (rough approximation)
    cost = sum([x['cost'] for x in lm.history if x['cost'] is not None])  # cost in USD, as calculated by LiteLLM for certain providers

    # Log results
    logger.info(
        "Optimization results",
        baseline_score=f"{baseline_score:.3f}",
        final_score=f"{final_score:.3f}",
        improvement=f"{improvement:.3f}",
        duration=f"{duration:.2f}s",
        estimated_cost_usd=cost,
    )

    # Save results
    results = {
        "baseline_score": baseline_score,
        "final_score": final_score,
        "improvement": improvement,
        "duration": duration,
        "estimated_cost_usd": cost,
    }

    # Ensure results directory exists
    Path("optimizers/results").mkdir(parents=True, exist_ok=True)

    with open("optimizers/results/optimization_results.json", "w", encoding="utf-8") as f:
        json.dump(results, f, indent=2, ensure_ascii=False)

if __name__ == "__main__":
    main()
